selfdrive/car/gm/carcontroller.py

```diff
@@ -89,18 +92,18 @@ def update(self, CC, CS):
     time_and_headlights_step = 10
     tt = self.frame * DT_CTRL
 
-    if self.frame % time_and_headlights_step == 0:
+    if CS.CP.openpilotLongitudinalControl and (not CS.CP.radarOffCan) and (self.frame % time_and_headlights_step) == 0:
       idx = (self.frame // time_and_headlights_step) % 4
       can_sends.append(gmcan.create_adas_time_status(CanBus.OBSTACLE, int((tt - self.start_time) * 60), idx))
       can_sends.append(gmcan.create_adas_headlights_status(self.packer_obj, CanBus.OBSTACLE))
 
     speed_and_accelerometer_step = 2
-    if self.frame % speed_and_accelerometer_step == 0:
+    if CS.CP.openpilotLongitudinalControl and (not CS.CP.radarOffCan) and self.frame % speed_and_accelerometer_step == 0:
       idx = (self.frame // speed_and_accelerometer_step) % 4
       can_sends.append(gmcan.create_adas_steering_status(CanBus.OBSTACLE, idx))
       can_sends.append(gmcan.create_adas_accelerometer_speed_status(CanBus.OBSTACLE, CS.out.vEgo, idx))
 
-    if self.frame % self.params.ADAS_KEEPALIVE_STEP == 0:
+    if CS.CP.openpilotLongitudinalControl and CS.CP.networkLocation == NetworkLocation.gateway and (self.frame % self.params.ADAS_KEEPALIVE_STEP) == 0:
```

Comment:
OP only needs to send keepalives when the factory ECU cannot. Presently only applies in the ASCM case because the entire ASCM is taken off the PT bus.